draft-ietf-privacypass-auth-scheme-extensions.md

```diff
@@ -123,17 +127,63 @@ issuance protocol. In particular, the value of an Extensions structure is provid
 metadata for the issuance protocol. Candidate issuance protocols are specified in
 {{?PUBLIC-ISSUANCE=I-D.hendrickson-privacypass-public-metadata}}.
 
-# Extensions Negotiation {#negotiation}
+# PrivateToken Challenge Extension Parameter
+
+As defined in {{Section 2.1 of AUTHSCHEME}}, the "PrivateToken" authentication
+scheme defines two parameters, "challenge" which contains the base64url-encoded
+TokenChallenge struct, and a "token-key" which contains the base64url-encoded
+public key used for this challenge. This document defines two OPTIONAL new parameters, 
+"extension-set," which contains the base64url-encoded representation of the
+following ExtensionSet structure, and "extensions" which contain the base64url-encoded 
+representation of the Extensions strucuture defined in {#extensions}. This document 
+follows the default padding behavior described in {{Section 3.2 of !RFC4648}}, so the
+base64url value MUST include padding.
+
+~~~
+struct {
+    enum { false(0), true(1) } Bool;
+    Bool is_required;
+    ExtensionType extension_type;
+} ExtensionEntry;
+
+enum {
+    reserved(0),
+    (65535)
+} ExtensionType;
+
+struct {
+    ExtensionEntry extension_types<0..2^16-1>;
+} ExtensionSet;
+~~~
+
+The contents of ExtensionSet is a list of ExtensionEntry structs containing extensions (defined in #extensions),
+each of which is a type-length-value structure whose semantics are determined by the type, and a bit marking whether
+the extension is required or optional. T he type and length of each ExtensionType is a 2-octet integer, in network byte order. The
+length of the extension_types list is also a 2-octet integer, in network byte order.
+
+ExtensionTypes are to be defined outside of this document.
 
-The mechanism Clients and Origins use to determine which set of extensions to provide
-for redemption is out of scope for this document.
+The extensions parameter is to be used for pre-populated extension structs the origin suggests
+to the client.
+
+When presented with an ExtensionSet, a client should expect to be rejected if not providing required extensions.
+A client MAY choose to provide optional extensions or not. A client can choose to use the pre-populated extension
+provided by the origin, or craft its own.
+
+~~~
+WWW-Authenticate:
+  PrivateToken challenge="abc...", token-key="123...", extension-set="0x0001,0x0002..." extensions="def..."
+~~~
+
+# Extensions Negotiation {#negotiation}
```

Comment:
How does client process these two new parameters? since "extension-set" and "extensions" carry ExtensionType  is there redundancy in these parameters? Can we instead just have "extensions", containing a list of Extension structs? Extension struct may carry an empty extension_data value, which indicates any or wild card extension parameter value.

Reply:
Would challenge, token-key and extension-set not suffice?